Add Survey Dialog on Plugins Deactivate

cbauerman opened the pull request:
#### Changes proposed in this Pull Request:

* Load a script with Jetpack so that we can intercept clicks on the plugins screen `/wp-admin/plugins.php`
* Show Jetpack Disconnect Modal warning window when user try to deactivate Jetpack

**Is this a new feature or does it add/remove features to an existing part of Jetpack?**
* Internal reference: parBdM-hs-p2

#### Testing instructions:
1. navigate to `/wp-admin/plugins.php` with Jetpack Active and connected
2. click "deactivate" for Jetpack
<img width="650" alt="Screen Shot 2020-03-12 at 3 31 03 PM" src="https://user-images.githubusercontent.com/2810519/76572335-fbb3bb00-6476-11ea-871d-dac744b8900a.png">
3. Verify the dialog works as it does on the dashboard
<img width="650" alt="Screen Shot 2020-03-12 at 3 31 34 PM" src="https://user-images.githubusercontent.com/2810519/76572452-5d742500-6477-11ea-8fb9-4ce80d48e177.png">

4. Verify that click "Cancel" cancels the deactivation

5. Verify that after clicking "Disconnect and Deactivate" you are taken back to the plugins screen with Jetpack deactivated.

6. Connect Jetpack again and visit Jetpack Dashboard

7. Click Manage Connection and check that the modal is still working as expected there

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Add Jetpack Disconnect Modal to WordPress Plugins screen deactivate link

cbauerman committed:
hook up survey to endpoint

cbauerman committed:
[not verified] hook up survey to endpoint

cbauerman committed:
[not verified] remove extra README file

cbauerman committed:
remove extra README file

cbauerman committed:
break apart long function

cbauerman committed:
[not verified] break apart long function

cbauerman committed:
[not verified] rename survey to disconnect-survey

cbauerman committed:
rename survey to disconnect-survey

cbauerman committed:
[not verified] add deactivate route

cbauerman committed:
add deactivate route

cbauerman committed:
don't disable cancel button in survey

cbauerman committed:
fix disconnect-survey path

cbauerman committed:
[not verified] fix disconnect-survey path

cbauerman committed:
[not verified] add rendering, working plugins dialog

cbauerman committed:
[not verified] [not verified] add rendering, working plugins dialog

cbauerman committed:
lock dialog once disconnect begins

jeherve commented:
This should be good to rebase against `master` now!

leogermani commented on class.jetpack.php:
```diff
@@ -5928,6 +5930,24 @@ function add_token_to_login_redirect_json_api_authorization( $redirect_to, $orig
+	function deactivate_dialog() {
+ 		if ( Jetpack::is_active() ) {
+ 			add_thickbox();
+ 			wp_enqueue_script(
```
Couldn't we add it only at the Plugins page?

leogermani replied:
done!

cbauerman commented:
closing this PR as the survey has been indefinitely postponed

leogermani commented:
Re-opening this PR.

In #15456 we added a new section to the disconnection modal, warning the user that other plugins are relying on the Jetpack connection.

Since other plugins are starting using the connection, we think it's important to add this warning so users don't break their sites by mistake.

We decided to reopen this PR and finish this so we can have the same warning when deactivating Jetpack.

leogermani committed:
Change label to something more explicit

leogermani commented:
I just refreshed this branch with latest `master` and tested it.

Also, changed the label of the confirmation button to "Disconnect and Deactivate" to be more explicit and match the question that comes before "are you sure you want to disconnect and deactivate?"

This still needs:

- [ ] Design. see #15768
- [ ] layout adjustments (scroll, paddings)

eeeeevon13 commented:
Let's model this modal (see what I did there?) after this one:
**Should look like this**
<img width="563" alt="Screenshot 2020-05-14 10 45 43" src="https://user-images.githubusercontent.com/49159284/81955865-3e595700-95d0-11ea-8c6a-1f4a65443eea.png">
**Changes to your pr**

1. Making heading regular font not bold to match the one pictured above
2. Stats number inside the box should be regular not bold
3. Pull the close button inside the modal (as pictured above)
4.  The subhead looks like it needs to be bumped up a bit -looks small

![modal](https://user-images.githubusercontent.com/49159284/81956051-7b254e00-95d0-11ea-9517-e1d6f4fc085b.jpg)

cbauerman commented:
> Let's model this modal (see what I did there?) after this one:
> **Should look like this**
> <img alt="Screenshot 2020-05-14 10 45 43" width="563" src="https://user-images.githubusercontent.com/49159284/81955865-3e595700-95d0-11ea-8c6a-1f4a65443eea.png">
> **Changes to your pr**
> 
> 1. Making heading regular font not bold, looks like it may need to be bigger as well, Can you pull the styles directly from the other modal? (pictured above)
> 2. Stats number inside the box should be regular not bold
> 3. Pull the close button inside the modal, make gray an larger, follow the other modal (as pictured above)
> 4. The subhead looks like it needs to be bumped up a bit -looks small
> 5. Is there  a huge margin at the bottom causing the scrollbar?
> 
> ![modal](https://user-images.githubusercontent.com/49159284/81956051-7b254e00-95d0-11ea-9517-e1d6f4fc085b.jpg)

For what it is worth that is my very old screenshot, many of those issues should be fixed!

leogermani commented:
> For what it is worth that is my very old screenshot, many of those issues should be fixed!

Yes, that's on me. I wasn't expecting a reply so fast! :) 

Anyway, I'm starting digging into the way you did to reuse the modal and for alternatives, to see if we can fix the close button, the scroll and bottom margin.

leogermani committed:
[not verified] Deactivate modal layout adjustments

Many adjustments to make it look as much as possible
with the modal on the Jetpack Dashboard

leogermani committed:
[not verified] adust title color

leogermani commented:
I've managed to make most of the layout adjustments, but there are still some limitations. Here is how it's looking like now:

![Captura de tela de 2020-05-14 19-40-09](https://user-images.githubusercontent.com/971483/81992822-cc5a2f80-961a-11ea-9bc1-260c1087b3b1.png)

We need to test and see if this is acceptable, or if we should go for a completely different approach, maybe refactoring how this modal is created and changing it in a way it can be reused in a better way.

## First, user experience is quite bad

The first thing is that when we click to Deactivate, the user sees a blank modal, then the default empty Admin background, with the Jetpack logo, and then everything gets overridden by a spinner... and only then the actual content is loaded.

Give it a try! (I tried to record a gif but it didn't capture all the stages)

## Window height

Thickbox auto calculates the window height, so we see there is unused space at the bottom of the window. I tried to set the body background to transparent, but the thickbox dropshadow then looks broken. 

Also, if we don't have some margin at the bottom, the buttons are hidden if the modal content gets too big and require a scrollbar. Give it a try inspecting and adding more content to force the scrollbar

## Title, fonts, etc...

The thickbox close button is outside the iframe, it won't scroll together with the iframe content. So I had to remove the title from the iframe/React content and use the Thickbox title.

If you look closely, you'll notice that the close button is different from the Modal on the Jetpack Dashboard and I can't really tell if the Title font is exactly with the same size and weight. It should be but it looks somewhat different to me.

TODO:
- [ ] adjust for small screens

cc @dereksmart

leogermani committed:
[not verified] copy and layout adjustmets on plugin list

leogermani committed:
[not verified] fix identation

leogermani commented:
After some discussion with @dereksmart, we decided to close this PR and go for another approach.

The main reason is that the user experience provided by this `iframe` approach is not quite good. Here's what happens when you click "Deactivate":

1. a blank modal pops up and keeps blank while loading the iframe content
![modal-1](https://user-images.githubusercontent.com/971483/82099363-929e2d00-96dd-11ea-810d-89ab6dbe763b.png)

2. The iframe loads, and shows the default empty Admin background, with the Jetpack logo at the top

![modal-2](https://user-images.githubusercontent.com/971483/82099361-92059680-96dd-11ea-9fe5-3e84476cd27e.png)

3. The default admin background is replaced by the Modal window with a spinner while it loads the modal content
![modal-3](https://user-images.githubusercontent.com/971483/82099360-916d0000-96dd-11ea-85bf-36f6bbb7d0bf.png)

4. Finally, it loads the modal content
![modal-4](https://user-images.githubusercontent.com/971483/82099358-90d46980-96dd-11ea-8ea4-a34e6b6e7082.png)

It also has a bottom margin that's hard to adjust and some other minor inconveniences.

Some of these problems could be minimized, but the overall experience would hardly be much better.

We have an urgent demand to show up a list of active plugins relying on the Jetpack connection there, to warn users and keep them to break their sites by accident. We might go for a simpler approach, with a simples modal there, while we discuss a more complete refactor of this feature.

Some discussion on this has been going on here: p6TEKc-3FC-p2

The work related to warning users about disconnecting other plugins is being tracked on this MT: p1HpG7-8VJ-p2
